src/nvim/eval/decode.c

```diff
@@ -965,9 +964,8 @@ int msgpack_to_vim(const msgpack_object mobj, typval_T *const rettv)
         }));
         uint64_t n = -((uint64_t) mobj.via.i64);
         list_append_number(list, -1);
-        list_append_number(list, (varnumber_T) ((n >> 62) & 0x3));
-        list_append_number(list, (varnumber_T) ((n >> 31) & 0x7FFFFFFF));
-        list_append_number(list, (varnumber_T) (n & 0x7FFFFFFF));
+        list_append_number(list, (varnumber_T)((n >> 63) & 0x1));
+        list_append_number(list, (varnumber_T)(n & 0x7fffffffffffffff));
```

Comment:
old shada files should only contain 32-bit msgpack values (where a typval is encoded/decoded), which will be interpreted the same. New shada files with 64-bit values will be missinterpreted by old nvim (values get truncated), of course. (maybe better an error, but, hard to change the code of nvim executables out in the wild...) But 32-bit values in new shada files will still be interpreted correctly by old nvim, so it is "as compatible it possibly could be".